File: NASA_JulesKouatchou/Problem1/copyMatrix.jl
# **********************************************************
# Given a nxnx3 matrix A, we want to perform the operations:
#
#       A[i][j][0] = A[i][j][1]
#       A[i][j][2] = A[i][j][0]
#       A[i][j][1] = A[i][j][2]
# **********************************************************

#########################################
# Get the dimension from the command line
#########################################
n, = size(ARGS)
if (n < 1)
   println("Usage: copyMatrix.jl dim")
   println("       --->      Please specify the dimension.")
   exit()
end


dim = parse(Int, ARGS[1])

# Loop
#-----
function foo1(A)
   N = size(A, 1)
   for j = 1:N, i = 1:N
       A[i,j,1] = A[i,j,2]
       A[i,j,3] = A[i,j,1]
       A[i,j,2] = A[i,j,3]
   end
end

println("-------------------------------")
println(@sprintf "Copy of matrix (loop) %d" dim)
println("-------------------------------")

A = randn(dim,dim,3)

tic()
foo1(A)
toc()

println(" ")

println("--------------------------")
println(@sprintf "Vectorized Copy of matrix %d" dim)
println("--------------------------")

A = randn(dim,dim,3)

tic()

A[:,:,1] = A[:,:,2]
A[:,:,3] = A[:,:,1]
A[:,:,2] = A[:,:,3]

toc()
println(" ")

File: NASA_JulesKouatchou/Problem1/copyMatrix.f90
! **********************************************************
! Given a nxnx3 matrix A, we want to perform the operations:
!
!       A[i][j][0] = A[i][j][1]
!       A[i][j][2] = A[i][j][0]
!       A[i][j][1] = A[i][j][2]
! **********************************************************
      program copyMatrix

      real :: r
      integer dim1, i, j, l
      double precision, ALLOCATABLE :: A(:, :, :)
!      real start, finish, tarray(2)
      real start1, finish1
      real start2, finish2
      CHARACTER(len=30) :: arg(1)

      !----------------------------------------
      ! Get the dimension from the command line
      !----------------------------------------
      call getarg(1,arg(1))
      read(arg(1), * ) dim1
     
      ALLOCATE( A(dim1, dim1, 3))

      !-----------------
      ! Loop formulation
      !-----------------
      call srand(86456)
      do i = 1, dim1
        do j = 1, dim1
           do l = 1, 3
              call random_number(r)
              A(i, j, l) = r
          enddo
        enddo
      enddo

      call cpu_time(start1)

      do j = 1, dim1
         do i = 1, dim1
            A(i, j, 1) = A(i,j,2)
            A(i, j, 3) = A(i,j,1)
            A(i, j, 2) = A(i,j,3)
         enddo
      enddo

      call cpu_time(finish1)
      print *,'Time Matrix Copy - ji loop: ', dim1, finish1-start1,' s'

      !--------------------------
      ! Vectorization formulation
      !--------------------------
      call srand(86456)
      do i = 1, dim1
        do j = 1, dim1
           do l = 1, 3
              call random_number(r)
              A(i, j, l) = r
          enddo
        enddo
      enddo

      call cpu_time(start2)

      A(:,:,1) = A(:,:,2)
      A(:,:,3) = A(:,:,1)
      A(:,:,2) = A(:,:,3)

      call cpu_time(finish2)
      print *,'Time Matrix Copy - vectorization ', dim1, finish2 - start2,' s'

      end program copyMatrix
